In [19]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as ppt
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from xgboost.sklearn import XGBClassifier
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
import pickle

pd.set_option('display.max_columns', 50)

In [2]:
df = pd.read_csv('train_keystroke.csv')

In [3]:
df.describe()

user  press-0   release-0      press-1    release-1      press-2  \
count  880.000000    880.0  880.000000   880.000000   880.000000   880.000000   
mean    55.500000      0.0   69.740909   192.504545   280.730682   378.078409   
std     31.771009      0.0   35.830579   118.281523   128.337586   243.485675   
min      1.000000      0.0    0.000000     0.000000     1.000000     1.000000   
25%     28.000000      0.0   50.000000   152.000000   232.000000   281.000000   
50%     55.500000      0.0   72.000000   192.000000   279.000000   358.000000   
75%     83.000000      0.0   94.000000   231.250000   328.000000   424.000000   
max    110.000000      0.0  241.000000  1287.000000  1343.000000  4333.000000   

         release-2      press-3    release-3      press-4    release-4  \
count   880.000000   880.000000   880.000000   880.000000   880.000000   
mean    460.992045   625.425000   719.445455   792.982955   878.523864   
std     246.432639   359.327185   362.141729   403.595652   405.840245   
min       1.000000     2.000000     4.000000    14.000000    83.000000   
25%     374.000000   463.000000   561.000000   592.750000   676.000000   
50%     436.000000   569.000000   663.000000   721.500000   807.000000   
75%     501.250000   706.500000   800.000000   888.500000   978.500000   
max    4453.000000  4823.000000  4890.000000  5191.000000  5263.000000   

           press-5    release-5      press-6    release-6      press-7  \
count   880.000000   880.000000   880.000000   880.000000   880.000000   
mean   1043.360227  1131.386364  1242.047727  1327.745455  1485.427273   
std     465.689141   467.088020   525.164691   527.872734   605.835537   
min     221.000000   318.000000   388.000000   470.000000   562.000000   
25%     795.500000   889.500000   950.750000  1037.000000  1130.000000   
50%     944.000000  1034.000000  1131.500000  1214.500000  1346.500000   
75%    1161.500000  1247.750000  1391.250000  1471.250000  1688.250000   
max    6239.000000  6377.000000  6497.000000  6624.000000  7039.000000   

         release-7      press-8    release-8      press-9    release-9  \
count   880.000000   880.000000   880.000000   880.000000   880.000000   
mean   1575.250000  1683.110227  1773.718182  1874.559091  1973.450000   
std     607.896069   654.999470   657.806567   724.772838   725.310795   
min     648.000000   667.000000   763.000000   793.000000   924.000000   
25%    1222.500000  1288.500000  1375.000000  1433.500000  1532.500000   
50%    1432.000000  1526.000000  1618.500000  1695.000000  1791.500000   
75%    1788.000000  1921.250000  2004.000000  2145.000000  2256.000000   
max    7136.000000  7377.000000  7471.000000  7663.000000  7753.000000   

          press-10   release-10     press-11   release-11     press-12  \
count   880.000000   880.000000   880.000000   880.000000   880.000000   
mean   2093.876136  2194.747727  2272.895455  2366.973864  2491.186364   
std     793.458930   792.913889   856.416953   856.341509   905.936922   
min     891.000000  1028.000000   955.000000  1055.000000  1124.000000   
25%    1589.750000  1708.000000  1741.750000  1834.250000  1930.500000   
50%    1896.000000  1994.000000  2059.500000  2151.000000  2255.500000   
75%    2390.250000  2474.500000  2585.750000  2687.250000  2823.250000   
max    8006.000000  8063.000000  8189.000000  8259.000000  8422.000000   

        release-12  
count   880.000000  
mean   2587.061364  
std     905.291522  
min    1188.000000  
25%    2034.500000  
50%    2344.000000  
75%    2905.750000  
max    8487.000000

from describe(): it appears that one/many of the 880 rows only wrote 12 characters not 13, and that came from 'press-0' 'release-0' 'press-1'.

as HT is 0 for the first character and started writing from second character.

In [4]:
df[(df['press-0'] == 0) & (df['release-0'] == 0)] # 33 rows

user  press-0  release-0  press-1  release-1  press-2  release-2  \
33      5        0          0       41        185      209        297   
34      5        0          0        0         78      127        198   
111    14        0          0        7        100      264        359   
117    15        0          0      129        184      315        385   
147    19        0          0       79         99      139        235   
148    19        0          0        0          1        1          8   
155    20        0          0       51        121      217        289   
173    22        0          0        0          1        1         49   
174    22        0          0       98        152      349        392   
175    22        0          0       99        137      320        387   
259    33        0          0        6         72      188        255   
295    37        0          0      384        456      624        696   
300    38        0          0       69        140      404        458   
332    42        0          0        1          1        1         33   
347    44        0          0      126        197      293        454   
362    46        0          0        1          1        1          2   
410    52        0          0      108        261      203        398   
413    52        0          0        7        125       75        242   
414    52        0          0       13        113      112        252   
466    59        0          0        7        159      103        235   
486    61        0          0        0         55      191        239   
540    68        0          0        1          1        1          2   
607    76        0          0       48        128      147        315   
613    77        0          0       42        112      218        282   
621    78        0          0        0         15      115        163   
644    81        0          0       82        150      223        302   
646    81        0          0        1          1        1          2   
647    81        0          0        1          1        1          1   
658    83        0          0       41        121      307        352   
660    83        0          0      120        209      353        422   
717    90        0          0        0         37        1        104   
868   109        0          0        0          1        1         58   
870   109        0          0        0          1      100        198   

     press-3  release-3  press-4  release-4  press-5  release-5  press-6  \
33       418        489      585        657      705        753      967   
34       295        366      455        550      672        766      839   
111      566        636      781        877     1046       1109     1165   
117      455        527      672        745      861        942     1001   
147      379        475      499        598     1222       1286     1404   
148      143        240      290        390      496        571      691   
155     3077       3226     3153       3249     3392       3465     3614   
173      145        233      245        338      501        527      605   
174      462        558      535        630      749        838      914   
175      858        956     1003       1106     1230       1285     1355   
259      399        477      523        593      711        807      808   
295      836        906     1024       1101     1687       1759     2204   
300      828        972     1133       1214     1790       1828     1958   
332      106        177      326        394      513        596      627   
347      533        614      693        765      861        933     1053   
362      182        264      466        518      678        738      846   
410      491        540      675        748      891        963     1083   
413      387        461      625        690      931       1002     1099   
414      497        563      723        795     1179       1284     1387   
46

33 records only wrote 12 character
also noticed 'press-X' has very low values and 'release-X' also ( values like V < 10), assuming this numbers are in millisecond, that comes to conclude that this isn't a human (robot/script) or copy-pasted those characters 

as the model is focused to select which user from test data, and not indentify if this user is human or robot.
then we can replace those low values in columns ( X < 1st quartile (25%) or a certain value), with mean/mode value for the same column. to complete our data

In [5]:
df[(df['press-1'] <= 20) & (df['press-2'] <= 30)] #  24 rows

user  press-0  release-0  press-1  release-1  press-2  release-2  \
134    17        0          3        1          3        3         19   
146    19        0          1        1          1        1        102   
148    19        0          0        0          1        1          8   
151    19        0          1        1          1        2          2   
173    22        0          0        0          1        1         49   
257    33        0          4        4          4       15        117   
260    33        0          1        1          1        1          2   
262    33        0          1        1          1       23        110   
263    33        0          1        1          1        2          2   
332    42        0          0        1          1        1         33   
362    46        0          0        1          1        1          2   
386    49        0          1        1          2       24        121   
387    49        0          1        0          1        1         76   
388    49        0          1        1          1        1         56   
540    68        0          0        1          1        1          2   
541    68        0          3        3          3        3         40   
542    68        0          1        1          1        1          2   
645    81        0          1        1          2        2          2   
646    81        0          0        1          1        1          2   
647    81        0          0        1          1        1          1   
717    90        0          0        0         37        1        104   
866   109        0          1        1          6        7          7   
867   109        0          1        1          1        2          2   
868   109        0          0        0          1        1         58   

     press-3  release-3  press-4  release-4  press-5  release-5  press-6  \
134      402        570      478        611      776        915     1022   
146      211        307      339        459      570        669      819   
148      143        240      290        390      496        571      691   
151        2          4       14         83      221        318      434   
173      145        233      245        338      501        527      605   
257      280        303      345        391      546        614      631   
260       86        159      253        349      589        686      780   
262      230        316      351        398      548        624      625   
263        2         52      132        174      299        388      388   
332      106        177      326        394      513        596      627   
362      182        264      466        518      678        738      846   
386      328        411      414        504      664        749      856   
387      241        348      354        444      562        658      876   
388      224        312      320        392      510        608      624   
540        7         63      180        282      748        825      844   
541       71        151      271        343      743        874      875   
542        2          8      152        239      320        440      416   
645       12        160      153        253      333        420      501   
646        9        117      118        212      332        404      498   
647       35        162      165        244      328        402      476   
717      165        298      265        346      499        638      697   
866        7         26       81        189      325        389      457   
867        2        110      265        360      531        648      684   
868      103        225      299        359      491        591      677   

     release-6  press-7  release-7  press-8  release-8  press-9  release-9  \
134       1090     1179       1275     1347       1541     1435       1555   
146        885     1140       1221     1337       1476     1522       1599   
148        786      954       1026     1151 

data needs to be treated carefully knowing it isn't clean data, and it will be clear what to do when we extract our features

# 1st Step - creating four new new features

In [6]:
df[f'hold time-00'] = df[f'release-0'] - df[f'press-0']
for i in range(1, 13):
    df[f'hold time-{i:02}'] = df[f'release-{i}'] - df[f'press-{i}']
    df[f'ppt-{i:02}'] = df[f'press-{i}'] - df[f'press-{i - 1}']
    df[f'rrt-{i:02}'] = df[f'release-{i}'] - df[f'release-{i - 1}']
    df[f'rpt-{i:02}'] = df[f'press-{i}'] - df[f'release-{i - 1}']
feature_cols = sorted(list(df.columns[27:]))
feature_cols.append('user')
feature_cols

['hold time-00',
 'hold time-01',
 'hold time-02',
 'hold time-03',
 'hold time-04',
 'hold time-05',
 'hold time-06',
 'hold time-07',
 'hold time-08',
 'hold time-09',
 'hold time-10',
 'hold time-11',
 'hold time-12',
 'ppt-01',
 'ppt-02',
 'ppt-03',
 'ppt-04',
 'ppt-05',
 'ppt-06',
 'ppt-07',
 'ppt-08',
 'ppt-09',
 'ppt-10',
 'ppt-11',
 'ppt-12',
 'rpt-01',
 'rpt-02',
 'rpt-03',
 'rpt-04',
 'rpt-05',
 'rpt-06',
 'rpt-07',
 'rpt-08',
 'rpt-09',
 'rpt-10',
 'rpt-11',
 'rpt-12',
 'rrt-01',
 'rrt-02',
 'rrt-03',
 'rrt-04',
 'rrt-05',
 'rrt-06',
 'rrt-07',
 'rrt-08',
 'rrt-09',
 'rrt-10',
 'rrt-11',
 'rrt-12',
 'user']

now df has (13 hold time - 12 ppt - 12 rrt - 12 rpt) new columns
and lets create a new df which has those new columns and analysis what we have nw

In [7]:
df

user  press-0  release-0  press-1  release-1  press-2  release-2  \
0       1        0        120      216        312      424        496   
1       1        0         95      168        265      360        455   
2       1        0         71      143        231      783        903   
3       1        0         95      144        263      353        431   
4       1        0         70      166        238      310        406   
..    ...      ...        ...      ...        ...      ...        ...   
875   110        0         80      225        306      444        465   
876   110        0         81      217        289      387        473   
877   110        0         79      216        300      393        519   
878   110        0         84      200        290      384        476   
879   110        0         64      271        366      424        510   

     press-3  release-3  press-4  release-4  press-5  release-5  press-6  \
0        592        664      808        856     1000       1072     1304   
1        527        599      736        807      928        999     1024   
2       1087       1159     1351       1454     1559       1631     1703   
3        760        832     1159       1207     1327       1377     1500   
4        526        598      710        758      878        950      926   
..       ...        ...      ...        ...      ...        ...      ...   
875      575        672      689        784      884       1040     1000   
876      954       1057     1091       1189     1265       1419     1384   
877      692        807      807        894     1004       1204     1161   
878      782        930      943        978     1064       1240     1200   
879      703        798      827        928     1032       1200     1158   

     release-6  press-7  release-7  press-8  release-8  press-9  release-9  \
0         1400     1496       1544     1712       1760     1992       2064   
1         1095     1215       1271     1423       1471     1664       1711   
2         1799     1823       1902     2039       2111     2271       2343   
3         1591     2968       3015     3151       3223     3415       3463   
4         1022     1094       1166     1310       1382     1543       1605   
..         ...      ...        ...      ...        ...      ...        ...   
875       1123     1439       1576     1638       1735     1785       1884   
876       1465     1595       1698     1755       1866     1921       2019   
877       1279     1407       1512     1591       1724     1763       1866   
878       1310     1413       1484     1537       1658     1760       1848   
879       1299     1345       1445     1561       1624     1679       1798   

     press-10  release-10  press-11  release-11  ...  rpt-06  hold time-07  \
0        2376        2448      2584        2632  ...     232            48   
1        1880        1952      2039        2111  ...      25            56   
2        2487        2559      2679        2751  ...      72            79   
3        3631        3703      3815        3887  ...     123            47   
4        1734        1806      1926        1998  ...     -24            72   
..        ...         ...       ...         ...  ...     ...           ...   
875      1928        2040      2057        2186  ...     -40           137   
876      2161        2265      2351        2457  ...     -35           103   
877      1879        2046      2048        2192  ...     -43           105   
878      1831        1940      1944        2082  ...     -40            71   
879      1838        1966      1991        2096  ...     -42           100   

     ppt-07  rrt-07  rpt-07  hold time-08  ppt-08  rrt-08  rpt-08  \
0       192     144      96            48     216     216     168   
1       191     176     120            48     208     200     152   
2       120     103      24            72     216     209     137   
3      1468    1424    1377            72     183     208     136

In [8]:
ndf = df[feature_cols].copy()
ndf.reset_index(inplace=True)

In [9]:
ndf.set_index('index', inplace=True)

In [10]:
ndf.describe()

hold time-00  hold time-01  hold time-02  hold time-03  hold time-04  \
count    880.000000    880.000000    880.000000    880.000000    880.000000   
mean      69.740909     88.226136     82.913636     94.020455     85.540909   
std       35.830579     39.349084     31.485219     36.921767     30.417956   
min        0.000000      0.000000      0.000000      1.000000      1.000000   
25%       50.000000     68.000000     67.000000     71.000000     68.000000   
50%       72.000000     82.000000     76.000000     91.500000     82.000000   
75%       94.000000    103.250000     97.000000    118.000000    100.000000   
max      241.000000    289.000000    264.000000    265.000000    273.000000   

       hold time-05  hold time-06  hold time-07  hold time-08  hold time-09  \
count    880.000000    880.000000    880.000000    880.000000    880.000000   
mean      88.026136     85.697727     89.822727     90.607955     98.890909   
std       30.164517     31.290914     30.160706     38.364721     36.269305   
min        1.000000      1.000000      1.000000      1.000000      4.000000   
25%       70.000000     66.750000     71.000000     68.000000     72.000000   
50%       87.500000     81.000000     90.000000     83.000000     96.000000   
75%      105.000000    102.000000    105.000000    110.000000    120.000000   
max      215.000000    257.000000    250.000000    338.000000    294.000000   

       hold time-10  hold time-11  hold time-12       ppt-01       ppt-02  \
count    880.000000    880.000000    880.000000   880.000000   880.000000   
mean     100.871591     94.078409     95.875000   192.504545   185.573864   
std       43.352964     41.590008     33.667082   118.281523   194.601764   
min        0.000000      1.000000     15.000000     0.000000     0.000000   
25%       72.000000     70.000000     72.000000   152.000000   125.000000   
50%       94.500000     88.000000     95.000000   192.000000   168.000000   
75%      124.000000    111.000000    116.000000   231.250000   203.000000   
max      303.000000    348.000000    239.000000  1287.000000  4031.000000   

            ppt-03       ppt-04       ppt-05       ppt-06       ppt-07  \
count   880.000000   880.000000   880.000000   880.000000   880.000000   
mean    247.346591   167.557955   250.377273   198.687500   243.379545   
std     222.058021   119.499464   157.805531   159.435206   204.335014   
min       0.000000     4.000000     7.000000    23.000000    30.000000   
25%     147.000000    97.000000   185.750000   126.000000   144.000000   
50%     216.000000   145.000000   216.000000   166.000000   181.500000   
75%     289.250000   204.500000   264.250000   210.000000   263.250000   
max    4255.000000  1477.000000  3111.000000  1703.000000  2443.000000   

            ppt-08       ppt-09       ppt-10       ppt-11       ppt-12  \
count   880.000000   880.000000   880.000000   880.000000   880.000000   
mean    197.682955   191.448864   219.317045   179.019318   218.290909   
std     106.241426   169.253180   181.904564   144.299929   123.585188   
min      16.000000    24.000000    24.000000     1.000000    13.000000   
25%     142.000000   118.000000   136.000000   115.000000   169.000000   
50%     185.000000   155.000000   180.000000   152.000000   202.000000   
75%     228.250000   208.000000   240.000000   208.000000   240.000000   
max    1850.000000  2257.000000  2133.000000  2208.000000  1856.000000   

            rpt-01       rpt-02       rpt-03       rpt-04       rpt-05  \
count   880.000000   880.000000   880.000000   880.000000   880.000000   
mean    122.763636    97.347727   164.432955    73.537500   164.836364   
std     112.888494   196.557997   222.402351   125.062209   155.331480   
min    -117.000000   -97.000000   -89.000000  -119.000000   -98.000000   
25%      84.750000    22.000000    69.000000    -1.000000   107.000000   
50%     121.000000    88.000000   135.500000    64.000000   135.000000   
75%     152.250000   130.000000   209

from only min for hold time, you can see that a lot of characters have been automated ( written by script or copy pasted)

also noticed, negative values in rpt-X, which could be typing fast where pressing is faster than releasing. (release > press), and couldn't neglect the sign because this will change the data and produced inaccurate data.

# 2nd Step - creating our training features

calculating the mean and std for each one of 4 main features and from that a new data frame which will have the final 8 features and user class

In [11]:
ndf['hold time-mean'] = ndf[feature_cols[:13]].mean(axis=1)
ndf['ppt-mean'] = ndf[feature_cols[13:25]].mean(axis=1)
ndf['rpt-mean'] = ndf[feature_cols[25:37]].mean(axis=1)
ndf['rrt-mean'] = ndf[feature_cols[37:49]].mean(axis=1)
ndf['hold time-std'] = ndf[feature_cols[:13]].std(axis=1)
ndf['ppt-std'] = ndf[feature_cols[13:25]].std(axis=1)
ndf['rpt-std'] = ndf[feature_cols[25:37]].std(axis=1)
ndf['rrt-std'] = ndf[feature_cols[37:49]].std(axis=1)

In [12]:
ndf[['hold time-mean', 'ppt-mean', 'rpt-mean', 'rrt-mean', 'hold time-std', 'ppt-std', 'rpt-std', 'rrt-std', 'user']]

hold time-mean    ppt-mean    rpt-mean    rrt-mean  hold time-std  \
index                                                                      
0           72.000000  229.333333  157.333333  225.333333      21.908902   
1           70.384615  185.916667  113.666667  182.000000      17.523976   
2           80.307692  239.250000  156.833333  237.916667      17.172578   
3           72.000000  331.916667  259.916667  330.000000      21.205345   
4           74.769231  173.833333  100.833333  176.000000      13.869354   
...               ...         ...         ...         ...            ...   
875        111.307692  190.333333   88.083333  202.000000      46.370581   
876        103.076923  213.416667  113.666667  218.583333      23.350452   
877        123.615385  191.916667   70.166667  197.500000      34.767174   
878        108.230769  181.166667   76.000000  186.250000      37.550308   
879        109.538462  183.250000   77.833333  191.166667      32.513114   

          ppt-std     rpt-std     rrt-std  user  
index                                            
0       63.115094   67.910815   73.069498     1  
1       36.412931   46.529235   40.174619     1  
2      136.361235  133.531633  145.322499     1  
3      368.330498  362.565720  353.436377     1  
4       47.510445   47.613660   39.874804     1  
...           ...         ...         ...   ...  
875     89.469480   89.940848  101.799268   110  
876    116.619395  125.816124  128.481664   110  
877     54.841024   66.644011   73.933016   110  
878     84.555343   89.631569  106.248529   110  
879     54.359285   71.106621   70.345683   110  

[880 rows x 9 columns]

assigning features to variable X and user class to variable y and splitting the data to train dataand test data

In [22]:
X = ndf[['hold time-mean', 'ppt-mean', 'rpt-mean', 'rrt-mean', 'hold time-std', 'ppt-std', 'rpt-std', 'rrt-std']].copy()
y = df['user'].copy()
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# 3rd Step - Training machine learning models

creating pipelines for randomforest, support vector machine, and xgboost. As specified, no need for feature selection, cross valudation or parameters tuning, the classifier is using the default values.

In [43]:
rf = Pipeline([
    ('cls', RandomForestClassifier())
    ])
rf.fit(X_train, y_train)
rf.score(X_test, y_test)

0.19886363636363635

In [44]:
svm = Pipeline([
    ('cls', SVC())
])
svm.fit(X_train, y_train)
svm.score(X_test, y_test)

0.028409090909090908

In [45]:
xgb = Pipeline([
    ('cls', XGBClassifier())
])

xgb.fit(X_train, y_train)
xgb.score(X_test, y_test)

C:\Users\Muhammad\AppData\Roaming\Python\Python38\site-packages\xgboost\sklearn.py:1224: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


[19:41:36] WARNING: D:\Build\xgboost\xgboost-1.5.1.git\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'multi:softprob' was changed from 'merror' to 'mlogloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


0.21022727272727273

In [42]:
pickle.dump(rf, open('rf.model', 'wb'))
pickle.dump(svm, open('svm.model', 'wb'))
pickle.dump(xgb, open('xgb.model', 'wb'))